lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -891,95 +875,44 @@ class ADContext {
-  }
-
-  /// Tries to promote origFnOperand which is of function type to an autodiff
-  /// function.
+private:
+  /// Promotes the given `autodiff_function` instruction to a valid
+  /// `@differentiable` function-typed value.
   SILValue promoteToDifferentiableFunction(
-      SILBuilder &builder, SILLocation loc, SILValue origFnOperand,
-      AutoDiffIndexSubset *parameterIndices,
-      unsigned differentiationOrder, DifferentiationInvoker invoker);
+      AutoDiffFunctionInst *inst, SILBuilder &builder, SILLocation loc,
+      DifferentiationInvoker invoker);
 
-  /// For an autodiff_function instruction that is missing associated functions,
-  /// fill those in.
+public:
+  /// Process the given `[differentiable]` attribute, filling in JVP/VJPs if
+  /// missing.
+  bool processDifferentiableAttribute(
+      SILFunction *original, SILDifferentiableAttr *attr,
+      DifferentiationInvoker invoker);
+
+  /// Process the given `autodiff_function` instruction, filling in missing
+  /// associated functions if necessary.
   bool processAutoDiffFunctionInst(AutoDiffFunctionInst *adfi);
 
-public:
-  /// Register a differentiation task in the global worklist. This will ensure
-  /// that a `[differentiable]` attribute will be generated for the specified
-  /// indices, and that primal/adjoint synthesis will be run in the
-  /// Differentiation pass.
-  ///
-  /// The function must either be a definition or be serialized.
-  DifferentiationTask *
-  registerDifferentiationTask(SILFunction *original,
-                              const SILAutoDiffIndices &indices,
-                              DifferentiationInvoker invoker) {
-    // Make sure this pair of original and indices is unique.
-    assert(!lookUpDifferentiationTask(original, indices));
-    ArrayRef<Requirement> contextualRequirements;
-    if (invoker.getKind() ==
-        DifferentiationInvoker::Kind::IndirectDifferentiation)
-      contextualRequirements = invoker.getIndirectDifferentiation()
-                                   .second->getAttribute()
-                                   ->getRequirements();
-    auto *attr = getOrCreateDifferentiableAttr(
-        original, indices, contextualRequirements);
-    std::unique_ptr<DifferentiationTask> task(
-        new DifferentiationTask(*this, original, std::move(attr), invoker));
-    differentiationTasks.push_back(std::move(task));
-    enqueuedTaskIndices.insert(
-        {{original, indices}, differentiationTasks.size() - 1});
-    return differentiationTasks.back().get();
-  }
+  /// Get or create an associated function index subset thunk from
+  /// `actualIndices` to `desiredIndices` for the given associated function
+  /// value and original function operand.
+  /// Calls `getOrCreateLinearMapIndexSubsetThunk` to thunk the linear map
+  /// returned by the associated function.
+  std::pair<SILFunction *, SubstitutionMap>
+  getOrCreateAssociatedFunctionIndexSubsetThunk(
```

Comment:
In SIL differentiation, "index subset" can mean both parameter indices and result indices. This name is a bit unclear to me.

```suggestion
  getOrCreateSubsetParametersThunkForAssociatedFunction(
```

Reply:
Thanks! `IndexSubset` was definitely a suboptimal name.